File: Projeto_7/tests/test_Filtro.cpp
#include <cassert>
#include "Filtro.h"
#include "Aluno.h"

void testFiltro() {
    Filtro<Aluno> filtro;
    filtro.adicionarElemento(Aluno("João", 8.5));
    filtro.adicionarElemento(Aluno("Maria", 6.5));
    filtro.adicionarElemento(Aluno("Pedro", 7.5));

    auto alunosAprovados = filtro.filtrarPorCondicao([](const Aluno& aluno) {
        return aluno.getMedia() >= 7.0;
    });

    assert(alunosAprovados.size() == 2);
    assert(alunosAprovados[0].getNome() == "João");
    assert(alunosAprovados[1].getNome() == "Pedro");
}

int main() {
    testFiltro();
    std::cout << "Todos os testes passaram!\n";
    return 0;
}

File: Projeto_7/tests/test_filtro.py
import unittest
from aluno import Aluno
from filtro import Filtro

class TestFiltro(unittest.TestCase):
    def test_filtrar_por_condicao(self):
        aluno1 = Aluno("Kaique", 1, "Engenharia", [7.5, 8.0, 6.5])
        aluno2 = Aluno("Maria", 2, "Medicina", [9.0, 8.5, 7.5])
        filtro = Filtro[Aluno]()
        filtro.adicionar_elemento(aluno1)
        filtro.adicionar_elemento(aluno2)

        alunos_filtrados = filtro.filtrar_por_condicao(lambda aluno: aluno.media() > 7.0)

        self.assertEqual(len(alunos_filtrados), 2)
        self.assertEqual(alunos_filtrados[0].nome, "Kaique")
        self.assertEqual(alunos_filtrados[1].nome, "Maria")

    def test_imprimir_todos(self):
        aluno1 = Aluno("Kaique", 1, "Engenharia", [7.5, 8.0, 6.5])
        filtro = Filtro[Aluno]()
        filtro.adicionar_elemento(aluno1)

        def acao(aluno):
            self.assertEqual(aluno.nome, "Kaique")
        
        filtro.imprimir_todos(acao)

if __name__ == "__main__":
    unittest.main()
